### DISCUSSION TITLE
Proposal: Account Activation Fund

### DISCUSSION THREAD

**Discussion by shefi** [XRPLF/XRPL-Standards#198] [XLS Ideas (pre standard proposal)]
# 1. Abstract

An XRP fund for account activation would make the user experience for XRPL wallets/interfaces/apps frictionless.

# 2. Specification
*(Don't know if this is currently or can be made possible technically. Someone with deep knowledge of the codebase can give an input on this.)*

# 3. Appendix

### A.1: Who contributes to the fund?

All funds currently getting burned on `AccountDelete` should be transferred to this fund. Also, all reserve XRPs that were sponsored by this fund, for those accounts activated this way, must be returned to this fund.

### A.2: Where will this fund live?

*(Don't know if this is currently or can be made possible technically. Someone with deep knowledge of the codebase can give an input on this.)*

### A.3: How will wallets/interfaces/apps draw from this fund to activate accounts?

*(Don't know if this is currently or can be made possible technically. Someone with deep knowledge of the codebase can give an input on this.)*

### A.4: Will it affect the deflationary nature of XRP?
Yes and no.

Yes, because the XRP on `AccountDelete` will not be completely burned as before.

No, because the XRP in this fund will only be used for the sole purpose of account activation and nothing else. Furthermore, common sense dictates that with less account activation friction, the usage of XRPL may be increased, therefore providing for higher deflationary effect over time through transaction fees.

### A.5: Will it protect against excessive account creation spam?

No, but the *attack vector will be limited* up to the maximum amount of the XRPs in the fund.

While the attack vector is not completely eliminated, it is minimized/limited. It can be quantified if storage planning needs to be done to be prepared for it. Thus, comparatively with removing the reserve requirement entirely (or setting it 0/low) this way of account activation sponsorships does have some protection in place.

### A.6: Can wallets/interfaces/apps abuse/drain the fund by hoarding activated accounts?

Yes.

However, the expectation would be that they will be playing fairly with it. There may be needed other ways to protect against this abuse if it's possible technically. Possibly throttle the amount of accounts they can activate within a specified period of time. Another way would be to force the return of the sponsored XRPs to the fund on their next transaction, as soon as the account has an amount of XRP >= to the amount provided by the fund for account activation.

**Comment by shefi** [comment]
I am just sharing this idea I expressed in one of the X spaces with @xVet and @xrpl365, where the friction of the reserve requirements was discussed.

Bare in mind, I have no knowledge of codebase, so what I am suggesting might not be possible technically. I may have not used the right terms to describe things either.

Someone with deep knowledge of the codebase should go through the idea and point out if it can be done or can't be done - technically or otherwise. I would appreciate it if they then can take over the specification proposal or work with me in completing the proposal specification.

Thanks.

**Comment by mvadari** [comment]
You currently have to sacrifice at least 2 XRP if you create a new account, since that's the cost of the `AccountDelete` transaction.

IMO this proposal ruins that financial incentive to not spam the ledger by creating tons of accounts, because you could recoup your losses (and even gain more, if you create your wallet from that pool). It actually might _incentivize_ creating and then quickly deleting accounts, since that's a free 8 XRP that you can transfer to any other wallet.

**Reply by shefi** [reply]
No, that's _not possible to gain_, if it's possible to implement as I have described it.

> "Also, all reserve XRPs that were sponsored by this fund, for those accounts activated this way, must be returned to this fund."

Meaning, if the account creation was sponsored by this fund/pool nothing will go to the wallet holder. It must all go back to the fund/pool.

They cannot gain anything this way.

They can however spam the ledger, like you said. However, as I have explained it in 3.1.A.5, it's a limited attack vector. Compared to the limitless attack vector, if no reserves are required.

**Reply by mvadari** [reply]
> "Also, all reserve XRPs that were sponsored by this fund, for those accounts activated this way, must be returned to this fund."

How would you handle the 8 XRP difference between the amount currently destroyed in `AccountDelete` and the amount needed for account reserve? Note that you can also burn those 8 XRP in fees.

**Reply by shefi** [reply]
If I understand your proposal correctly, ["0068 XLS-68d: Sponsored Fees and Reserves"](https://github.com/XRPLF/XRPL-Standards/discussions/196) it (or something similar) might be used for this purpose. So, instead of sponsored XRP coming from a sponsor, it comes from this fund/pool. When the account gets deleted, all the sponsored funds get returned to this fund/pool.

I don't know if/how this can be achieved technically in rippled, because like I said in my first comment, I don't have knowledge of the codebase.

I am just talking about the idea that, if the XRPs come from this fund/pool the whole amount that was drawn from it initially must be returned when account gets deleted.

**Reply by mvadari** [reply]
XLS-68d involves having a third party vet the sponsee to ensure that they're not trying to pull anything, and involves that party being the loaner. Having the network as a whole "loan" XRP to someone requires a separate level of trust that doesn't feel very feasible to me, since no one can truly be vetted to the network (there's always a point of trust).

It'll also be very difficult to keep this pool funded, since the `AccountDelete` fee is so much lower than the account reserve (and it is very unlikely that those will ever be equal). And no one will willingly donate to the pool.

**Reply by JosepMariaVila** [reply]
> Meaning, if the account creation was sponsored by this fund/pool nothing will go to the wallet holder. It must all go back to the fund/pool.
> 
If someone owns the private key for an account, and we want that the account is entirely an XRP account completely decentralized as any other account on the XRPL as they are right now, it doesn't seem possible to stop someone that owns the account to gain the 8 XRP by deleting the account and exploiting that repeatedly.

**Reply by shefi** [reply]
They don't have to be vetted, you can assume they may be malicious. However, they will not be able to do unlimited attacks. Worst case they create many accounts and lock that XRP away forever. Which is nothing different from what's happening now.

Regarding the amount of funds that can be collected from `AccountDelete`s, whatever amount will make a difference.

In addition, if this can also be implemented, the funding issue gets improved a lot and probably gets solved.

> Another way would be to force the return of the sponsored XRPs to the fund on their next transaction, as soon as the account has an amount of XRP >= to the amount provided by the fund for account activation.

In other words, the fund/pool gets replenished from:
- non-funded `AccountDelete` => 2 XRP 
- funded `AccountDelete` => 10 XRP 
- funded account reaches min threshold of >= funded amount, on next transaction, => 10 XRP 

This would solve the fund/pool growth feasibility issue.

**Reply by shefi** [reply]
@JosepMariaVila 

> If someone owns the private key for an account, and we want that the account is entirely an XRP account completely decentralized as any other account on the XRPL as they are right now, it doesn't seem possible to stop someone that owns the account to gain the 8 XRP by deleting the account and exploiting that repeatedly. 

Yeah, that I don't know.

That's why I asked for anyone with the needed knowledge of the codebase to look at it and give some input if/how it can be done technically. If it's currently not possible with what's there, maybe someone should look into if it can be done without degrading the performance/security/etc requirements.

**Reply by mvadari** [reply]
> They don't have to be vetted, you can assume they may be malicious. However, they will not be able to do unlimited attacks.

Why not? Anyone can keep creating an infinite number of accounts on the XRPL, provided they have a way to pay for the reserve.


> Worst case they create many accounts and lock that XRP away forever. Which is nothing different from what's happening now.

That's fine right now, when that XRP comes from a specific person. But when it's coming from the network itself, it becomes a much bigger deal. You get into discussions like who is "worthy" of that 

> whatever amount will make a difference. In addition, if this can also be implemented, the funding issue gets improved a lot and probably gets solved.

However much it is, it won't be enough. Now you have to deal with people who got accounts for free vs. people who weren't fast enough and had to pay for accounts. How do you decide what's "fair"? Is first come first serve really "fair", when that's definitely going to be gamed?

The reason fairness and gameability aren't a problem with XLS-68d is because it's not the network's problem or responsibility to handle that, it's the sponsor's (the proposal also doesn't involve the actual transfer of any XRP). They can make whatever rules they want. But we have to be fair when it's the network making the rules for what is essentially free XRP.

**Reply by shefi** [reply]
> Anyone can keep creating an infinite number of accounts on the XRPL, provided they have a way to pay for the reserve.

It's not infinite. However big the number is, it's quantifiable up to the max number of XRPs in the fund/pool.

> However much it is, it won't be enough.

I agree, it's not the perfect solution. However, compared to what's available right now, it does improve considerably on that pain point, if not eleminate it.

> How do you decide what's "fair"?

The question of fairness is very simple. Are the rules of the game known by all participants? Do the participants have the same access to the opportunity provided? There will never be an absolute fair system. There have to be some trade-offs on some level. First come first serve, on this sense is not an unfair mechanism.

> Now you have to deal with people who got accounts for free vs. people who weren't fast enough and had to pay for accounts.

There is nothing to deal with. If people created accounts with the available funds, that's the intended purpose. Those who missed out can wait for the funds to become available again to use this "free" option, or they pony up some funds of their own if they want to play now.

> fairness and gameability aren't a problem with XLS-68d

I am not suggesting these proposals have to be either or. They can be both. While the available funds will help as much as they can, for all other cases the sponsors will have to pony up some funds.

While there may be some gameability by some wallets/interfaces/apps the expectation is that they will play as fair as they can for their own good. Otherwise, if they are abusing this public good they will take some reputation risks and may get called out by other participants.

Either way, the existing alternative is a guaranteed 0 help with reducing the onboarding friction. This new alternative will help at some level greater than 0. Even if it's 1% better, it might still be considerable to provide some more use case opportunities for XRPL.

**Reply by mvadari** [reply]
> However, compared to what's available right now, it does improve considerably on that pain point, if not eleminate it.

I disagree. IMO this will cause more problems than what we currently have, due to the division it will cause and the scammers it will incentivize.

> Do the participants have the same access to the opportunity provided? 

No, they do not. Legitimate participants aren't going to be sitting and waiting for the XRP to be available in the pool, if they want an account that badly they're probably willing to put up 10 XRP for it. Scammers, on the other hand, are perfectly willing and able to write a script that waits for enough XRP to be added to the pool and then immediately creates an account to take it all.

IMO any method that disproportionately benefits scammers over the general public isn't something that should be implemented.

> While there may be some gameability by some wallets/interfaces/apps the expectation is that they will play as fair as they can for their own good. Otherwise, if they are abusing this public good they will take some reputation risks and may get called out by other participants.

It is not up to the wallets. Anyone can create an account on the XRPL, you don't have to do so via a wallet.

Requiring people to go through a wallet or service of some sort to obtain this public good XRP means that the services have to be registered somehow. Then, how do you decide if a service is legitimate and can take from this pot and isn't some fake service a scammer has set up? If someone has to decide, then the network is no longer decentralized. If the UNL has to decide, then that's a much heavier burden on them that is not currently expected of them, and it's not their job to decide who is legitimate and who isn't.

> Even if it's 1% better, it might still be considerable to provide some more use case opportunities for XRPL.

I disagree that it's better.

Let's put this in real-world terms. You need $5 to open a bank account. Whenever someone closes a bank account (which is quite rare), the account owner gets $4 back, and the remaining $1 gets turned into a waiver for the bank and is left in an unsupervised pile outside of the bank. Anyone can grab a chunk of waivers to open as many accounts as they would like.

Functionally, this pile of money will always have $0 in it, because people are going to be watching the pile, and as soon as there's some money, someone is going to notice and take it all to open their account and then close it, thereby obtaining their free $4.

This now only benefits the scammers, because anytime someone wants to open a legitimate account, the pile will be empty, and they'll have to sit and compete with all the other thieves/scammers for the money as soon as it's added to the pile. They likely won't want to go through that effort, and they won't bother creating a new account.

P.S. Please don't take anything that I'm saying here personally. I'm glad that more members of the community are thinking about these problems and how to solve them, and engaging in public discussion about it. I just strongly disagree that this is the solution.

**Reply by xrpl365** [reply]
First I want to applaud the writer of this document for doing more than talk about this on X. It's great to have non-devs contribute ideas for consideration. I hope the not-so positive responses so far will not put you or anyone else off from doing this in the future.

As for the proposal itself, again I applaud you for taking time to think of an idea and propose it. I 100% agree with the problem you have identified, however I can see several drawbacks to the proposal and only a small gain because the proposal sadly doesn't solve the issue in its entirety, so for that reason I would not support further development of this idea.

That said, I do not believe this should be the end of the discussion; we do need to have this debate and we absolutely need to find a workable solution to this issue. I have personally come up against it as a developer and this week witnessed first hand the shear frustration other devs are experiencing when the reality of the account reserve meets with their go to market strategy for an otherwise good business idea.

Well done Shef, you should be congratulated and next space I'm on with you I will give this a shout out.

One final last point...
Just because @mvadari and I have kind of poured some cold water on this doesn't mean you HAVE to stop. This is a decentralised network and you are free to continue to evolve your idea if you can gain support elsewhere. For anyone reading this it's important to understand that no-one person or entity can stop you if you feel compelled to continue. However if I were you I would take the feedback onboard and try to evolve your idea and maybe your next iteration may solve more of the issue and gain more traction.

I would be interested to hear what @dangell7 thought about your proposal.

**Reply by shefi** [reply]
@mvadari 

I apologize for only finding the time now to reply to your feedback.

I am going to go through your points a bit in reverse and hopefully that clears some of them up without much expansion on my side. 

> Let's put this in real-world terms. You need $5 to open a bank account. Whenever someone closes a bank account (which is quite rare), the account owner gets $4 back, and the remaining $1 gets turned into a waiver for the bank and is left in an unsupervised pile outside of the bank. Anyone can grab a chunk of waivers to open as many accounts as they would like.
> 
> Functionally, this pile of money will always have $0 in it, because people are going to be watching the pile, and as soon as there's some money, someone is going to notice and take it all to open their account and then close it, thereby obtaining their free $4.
> 
> This now only benefits the scammers, because anytime someone wants to open a legitimate account, the pile will be empty, and they'll have to sit and compete with all the other thieves/scammers for the money as soon as it's added to the pile. They likely won't want to go through that effort, and they won't bother creating a new account.

English is not my primary language, but what I am trying to say is if the money comes from the fund the account creator in your example gets $0 out of the whole process if done as I am describing it. If they used $ from the fund to open the account, when they close the account the whole amount must go back to the fund. So, when they close the account, if they used $ from the fund to open it, $1 + $4, total $5 gets returned to the fund.

I am suggesting for the same thing to happen with XRPL accounts. If they use XRP from the fund to open and activate the account, when they delete it 2 XRP burn + 8 XRP reserve = 10 XRP must go back to the fund. If/how this can be done is the question. The question is not to allow scammers get a portion of the XRP when they delete their account or not. That's out of the question. How can that be implemented/enforced technically is the thought and discussion we should have.


> I disagree. IMO this will cause more problems than what we currently have, due to the division it will cause and the scammers it will incentivize.

If implemented as I described above, there is no incentives for scammers. Other than the them hoarding activated accounts. They don't get any financial benefit from that, so they shouldn't have an incentive to do even that. 

> No, they do not. Legitimate participants aren't going to be sitting and waiting for the XRP to be available in the pool, if they want an account that badly they're probably willing to put up 10 XRP for it. Scammers, on the other hand, are perfectly willing and able to write a script that waits for enough XRP to be added to the pool and then immediately creates an account to take it all.

They wouldn't have to wait if proper maintenance is advocated by participants (wallets/apps etc), where users are educated to delete unused accounts. Over time the fund should keep growing to a level where this wouldn't be an issue. Maybe it can also be prefunded with funds from Ripple and/or the community so it picks up. As I said above, there must not be any incentive to those who would abuse it. No one should be able to get anything out of following this whole process.



> It is not up to the wallets. Anyone can create an account on the XRPL, you don't have to do so via a wallet.
> 
> Requiring people to go through a wallet or service of some sort to obtain this public good XRP means that the services have to be registered somehow. Then, how do you decide if a service is legitimate and can take from this pot and isn't some fake service a scammer has set up? If someone has to decide, then the network is no longer decentralized. If the UNL has to decide, then that's a much heavier burden on them that is not currently expected of them, and it's not their job to decide who is legitimate and who isn't.

You are right, it doesn't have to go through wallets. What I am describing should be implemented through the core APIs. I brought wallets as an example of those API consumers. It doesn't have to be wallet specific, but core ledger wide.

Nobody has to decide who can draw from the fund, anyone should be able to. The only risks involved in this participation should be that someone who abuses this would only be able to hoard activated accounts and they would be able to increase the storage for free up to the max that the XRPs of this fund would allow to. There should be no other benefit from this feature to those that would use it. I think these two tradeoffs would be minimalistic or calculated risks to have for the benefit this feature would provide.

@xrpl365 this update might help clear up a bit more on what I am trying to suggest with this change.

[a comment by an automated account is not shown]
